proxy/src/main/java/com/wavefront/agent/handlers/SourceTagSenderTask.java

```diff
@@ -0,0 +1,118 @@
+ * @author vasily@wavefront.com
+ */
+class SourceTagSenderTask extends AbstractSenderTask<SourceTag> {
+  private static final Logger logger =
+      Logger.getLogger(SourceTagSenderTask.class.getCanonicalName());
+
+  private final SourceTagAPI proxyAPI;
+  private final TaskQueue<SourceTagSubmissionTask> backlog;
+
+  /**
+   * Create new instance
+   *
+   * @param proxyAPI    handles interaction with Wavefront servers as well as queueing.
+   * @param handlerKey  metrics pipeline handler key.
+   * @param threadId    thread number.
+   * @param properties  container for mutable proxy settings.
+   * @param scheduler   executor service for this task
+   * @param backlog     backing queue
+   */
+  SourceTagSenderTask(HandlerKey handlerKey, SourceTagAPI proxyAPI,
+                      int threadId, EntityProperties properties,
+                      ScheduledExecutorService scheduler,
+                      TaskQueue<SourceTagSubmissionTask> backlog) {
+    super(handlerKey, threadId, properties, scheduler);
+    this.proxyAPI = proxyAPI;
+    this.backlog = backlog;
+  }
+
+  @Override
+  TaskResult processSingleBatch(List<SourceTag> batch) {
+    throw new UnsupportedOperationException("Not implemented");
+  }
+
+  @Override
+  public void run() {
+    long nextRunMillis = properties.getPushFlushInterval();
+    isSending = true;
+    try {
+      List<SourceTag> current = createBatch();
+      if (current.size() == 0) return;
+      Iterator<SourceTag> iterator = current.iterator();
+      while (iterator.hasNext()) {
+        if (rateLimiter == null || rateLimiter.tryAcquire()) {
+          SourceTag tag = iterator.next();
+          SourceTagSubmissionTask task = new SourceTagSubmissionTask(proxyAPI, properties,
+              backlog, handlerKey.getHandle(), tag, null);
+          TaskResult result = task.execute();
+          this.attemptedCounter.inc();
+          switch (result) {
+            case DELIVERED:
+              continue;
+            case PERSISTED:
+            case PERSISTED_RETRY:
+              if (rateLimiter != null) rateLimiter.recyclePermits(1);
+              continue;
+            case RETRY_LATER:
+              final List<SourceTag> remainingItems = new ArrayList<>();
+              remainingItems.add(tag);
+              iterator.forEachRemaining(remainingItems::add);
+              undoBatch(remainingItems);
+              if (rateLimiter != null) rateLimiter.recyclePermits(1);
+              return;
+            default:
+          }
+        } else {
+          final List<SourceTag> remainingItems = new ArrayList<>();
+          iterator.forEachRemaining(remainingItems::add);
+          undoBatch(remainingItems);
+          // if proxy rate limit exceeded, try again in 1/4..1/2 of flush interval
+          // to introduce some degree of fairness.
+          nextRunMillis = (int) (1 + Math.random()) * nextRunMillis / 4;
+          throttledLogger.info("[" + handlerKey.getHandle() + " thread " + threadId +
```

Comment:
in cases like this, can we use a function instead? maybe not for this CR but constructing these strings and evaluating Math.random() for something that could potentially be dropped due to throttling is just plain inefficient

Reply:
Fixed by switching to Supplier<String> instead.